src/niMidi.cpp

```diff
@@ -83,6 +95,71 @@ signed char convertSignedMidiValue(unsigned char value) {
 	return value - 128;
 }
 
+//-----------------------------------------------------------------------------------------------------------------------
+// The following conversion functions (C) 2006-2008 Cockos Incorporated
+// ToDo: Eliminate those that will eventually not be used (e.g. volToChar)
+static double charToVol(unsigned char val)
+{
+	double pos = ((double)val*1000.0) / 127.0;
+	pos = SLIDER2DB(pos);
+	return DB2VAL(pos);
+
+}
+
+static unsigned char volToChar(double vol)
+{
+	double d = (DB2SLIDER(VAL2DB(vol))*127.0 / 1000.0);
+	if (d < 0.0)d = 0.0;
+	else if (d > 127.0)d = 127.0;
+
+	return (unsigned char)(d + 0.5);
+}
+
+static double charToPan(unsigned char val)
+{
+	double pos = ((double)val*1000.0 + 0.5) / 127.0;
+
+	pos = (pos - 500.0) / 500.0;
+	if (fabs(pos) < 0.08) pos = 0.0;
+
+	return pos;
+}
+
+static unsigned char panToChar(double pan)
+{
+	pan = (pan + 1.0)*63.5;
+
+	if (pan < 0.0)pan = 0.0;
+	else if (pan > 127.0)pan = 127.0;
+
+	return (unsigned char)(pan + 0.5);
+}
+// End of conversion functions (C) 2006-2008 Cockos Incorporated
+//-----------------------------------------------------------------------------------------------------------------------
+
+static unsigned char peakToChar_KkMk2(double peak)
+{
+	// Direct linear logarithmic conversion to KK Mk2 Meter Scaling.
+	// Contrary to Reaper Peak Meters the dB interval spacing on KK Mk2 displays is NOT linear.
+	// It is assumed that other NI Keyboards use the same scaling for the meters.
+	// Midi #127 = +6dB #106 = 0dB, #68 = -12dB, #38 = -24dB, #16 = -48dB
+	double result = 0.0;
+	double a = -3.2391538612390192E+01;
+	double b = 3.0673618643561021E+01;
+	double c = 8.6720798984917224E+01;
+	double d = 4.4920143012996103E+00;
```

Comment:
What do these magic numbers represent? Please use named constants.

Reply:
they are the result of curve fitting, i.e. reverse engineering to fit Reaper's meters exactly to what is shown on the KK display. The KK display scale is a fixed pixel image. The function calculates the MIDI values that are required to show the bar at exactly the correct height to coincide with the dB value that Reaper reports. I.e. midi = a + b * ln(c * volume +d).
With these parameters the maximum error is 0.3% between Reaper values and display reading at the curve fitting points noted in the comments, i.e. far less then the quantization due to the pixel nature of the display.
![volToChar_kkMk2_FUNCTION](https://user-images.githubusercontent.com/48608100/58497303-86b60580-8130-11e9-98c4-e8fb4dcd2beb.png)

Reply:
I should add that this is the most CPU friendly function I found.